### PR TITLE
*: support update/delete/insert ignore/duplicate update in single cluster index table

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve?

Problem Summary:

support update/delete in single cluster index table

### What is changed and how it works?

Proposal: https://github.com/pingcap/tidb/blob/master/docs/design/2020-05-08-cluster-index.md

What's Changed: 

support update and delete single cluster index table

### Related changes

- n/a

### Check List 

Tests 

- Unit test

Side effects

- n/a

### Release note 

- No release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/17721)

### CODE REVIEW

File: executor/batch_checker.go [hunk 1]
```diff
@@ -106,21 +112,37 @@ func getKeysNeedCheckOneRow(ctx sessionctx.Context, t table.Table, row []types.D
...
+	var handleKey *keyValueWithDupInfo
+	if handle != nil {
 		handleKey = &keyValueWithDupInfo{
 			newKV: keyValue{
-				key:   t.RecordKey(kv.IntHandle(handle)),
+				key:   t.RecordKey(handle),
 				value: newRowValue,
 			},
-			dupErr: kv.ErrKeyExists.FastGenByArgs(strconv.FormatInt(handle, 10), "PRIMARY"),
+			dupErr: kv.ErrKeyExists.FastGenByArgs(handle.String(), "PRIMARY"),
```
Comment (coocood) [inline executor/batch_checker.go:145]: Use `stringutil.MemoizeStr` would reduce the String evaluate cost.

File: executor/update_test.go [hunk 2]
```diff
@@ -233,3 +233,62 @@ func (s *testUpdateSuite) TestUpdateMultiDatabaseTable(c *C) {
...
+	tk.MustExec(`create table ut3pk(id1 varchar(200), id2 varchar(200), v int, id3 int, primary key(id1, id2, id3))`)
+	tk.MustExec(`insert into ut3pk(id1, id2, v, id3) values ('aaa', 'bbb', 233, 111)`)
+	tk.MustQuery(`select id1, id2, id3, v from ut3pk where id1 = 'aaa' and id2 = 'bbb' and id3 = 111`).Check(testkit.Rows("aaa bbb 111 233"))
+	tk.MustExec(`update ut3pk set id1 = 'abc', id2 = 'bbb2', id3 = 222, v = 555 where id1 = 'aaa' and id2 = 'bbb' and id3 = 111`)
+	tk.MustQuery(`select id1, id2, id3, v from ut3pk where id1 = 'abc' and id2 = 'bbb2' and id3 = 222`).Check(testkit.Rows("abc bbb2 222 555"))
+	tk.MustQuery(`select id1, id2, id3, v from ut3pk`).Check(testkit.Rows("abc bbb2 222 555"))
+	tk.MustExec(`update ut3pk set v = 666 where id1 = 'abc' and id2 = 'bbb2' and id3 = 222`)
+	tk.MustQuery(`select id1, id2, id3, v from ut3pk`).Check(testkit.Rows("abc bbb2 222 666"))
+	tk.MustExec(`insert into ut3pk(id1, id2, id3, v) values ('abc', 'bbb3', 222, 777)`)
+	c.Assert(tk.ExecToErr(`update ut3pk set id2 = 'bbb3' where id1 = 'abc' and id2 = 'bbb2' and id3 = 222`).Error(), Equals, `[kv:1062]Duplicate entry '{abc, bbb3, 222}' for key 'PRIMARY'`)
```
Comment (coocood) [inline executor/update_test.go:271]: Need test case that updates only part of the primary key columns

File: table/tables/tables.go [hunk 3]
```diff
@@ -745,6 +751,23 @@ func DecodeRawRowData(ctx sessionctx.Context, meta *model.TableInfo, h kv.Handle
...
+			pkIdx := FindPrimaryIndex(meta)
+			var idxOfIdx int
+			for i, idxCol := range pkIdx.Columns {
+				if meta.Columns[idxCol.Offset].ID == col.ID {
+					idxOfIdx = i
+					break
+				}
+			}
+			dtBytes := h.EncodedCol(idxOfIdx)
+			_, dt, err := codec.DecodeOne(dtBytes)
```
Comment (coocood) [inline table/tables/tables.go:764]: Need field type to decode the column.

**Comment by coocood** [comment]
LGTM

**Comment by lysu** [comment]
WIP: found some bug about duplicate in the unique index and want to update pk - -

**Comment by lysu** [comment]
> WIP: found some bug about duplicate in the unique index and want to update pk - -

I will fire another PR to fix this(because DecodeHandleInUniqueIndexValue invoke many modifications- -), please help merge this PR if free @tangenta, thx~

File: executor/delete.go [hunk 4]
```diff
@@ -49,12 +50,28 @@ func (e *DeleteExec) Next(ctx context.Context, req *chunk.Chunk) error {
...
+			pkDts = append(pkDts, row[ordinal])
+		}
+		handleBytes, err := codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, nil, pkDts...)
+		if err != nil {
+			return err
+		}
+		handle, err = kv.NewCommonHandle(handleBytes)
+		if err != nil {
+			return err
+		}
```
Comment (tangenta) [inline executor/delete.go:73]: This seems duplicate with the lines in `getKeysNeedCheckOneRow`. Can we extract them into 
```
buildHandleFromIndexes(...) kv.Handle
```
?

File: executor/update_test.go [hunk 5]
```diff
@@ -233,3 +233,82 @@ func (s *testUpdateSuite) TestUpdateMultiDatabaseTable(c *C) {
...
+	tk.MustExec(`insert into t(id, v) values ('abc', 233)`)
+	tk.MustQuery(`select id, v from t where id = 'abc'`).Check(testkit.Rows("abc 233"))
+	tk.MustExec(`update t set id = 'dfg' where id = 'abc'`)
+	tk.MustQuery(`select * from t`).Check(testkit.Rows("dfg 233"))
+	tk.MustExec(`update t set id = 'aaa', v = 333 where id = 'dfg'`)
+	tk.MustQuery(`select * from t where id = 'aaa'`).Check(testkit.Rows("aaa 333"))
+	tk.MustExec(`update t set v = 222 where id = 'aaa'`)
+	tk.MustQuery(`select * from t where id = 'aaa'`).Check(testkit.Rows("aaa 222"))
+	tk.MustExec(`insert into t(id, v) values ('bbb', 111)`)
+	c.Assert(tk.ExecToErr(`update t set id = 'bbb' where id = 'aaa'`).Error(), Equals, `[kv:1062]Duplicate entry '{bbb}' for key 'PRIMARY'`)
```
Comment (tangenta) [inline executor/update_test.go:259]: How about using `tk.MustGetErrCode("xx", errno.ErrDupEntry)` to improve the flexibility for the future changes to error message?

File: executor/batch_checker.go [hunk 6]
```diff
@@ -76,28 +77,34 @@ func getKeysNeedCheck(ctx context.Context, sctx sessionctx.Context, t table.Tabl
...
-				handleCol = col
+				handleCols = append(handleCols, col)
 				break
 			}
 		}
 	}
+	if t.Meta().IsCommonHandle {
+		pkIdx := tables.FindPrimaryIndex(t.Meta())
+		for _, idxCol := range pkIdx.Columns {
+			handleCols = append(handleCols, t.Cols()[idxCol.Offset])
```
[the remark this answers is not in the document]
Comment (lysu) [reply executor/batch_checker.go:93]: IsCommonHandle should be false when PKIsHandle is true, so we cannot reuse 83 one

**Comment by lysu** [comment]
hi @coocood @tangenta, in last two commit, add some test case and handle about duplicate on unique key in cluster index table, please help relook last two commits, thanks~

**Comment by coocood** [comment]
LGTM

File: executor/batch_checker.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tangenta) [reply executor/batch_checker.go:93]: I mean we can `cols := t.Cols()` and reuse `cols`.

File: executor/delete.go [hunk 7]
```diff
@@ -119,16 +132,33 @@ func (e *DeleteExec) deleteSingleTableByChunk(ctx context.Context) error {
...
+				pkDts = append(pkDts, joinedRow[ordinal])
+			}
+			handleBytes, err := codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, nil, pkDts...)
+			if err != nil {
+				return err
+			}
+			handle, err = kv.NewCommonHandle(handleBytes)
+			if err != nil {
+				return err
+			}
```
Comment (tangenta) [inline executor/delete.go:156]:
```suggestion
			kv.BuildHandleFromDatumRow(e.ctx.GetSessionVars().StmtCtx, joinedRow, info.HandleOrdinal)
```

**Pull request merged**